pkg/kubelet/kubelet.go

```diff
@@ -905,14 +905,17 @@ func (kl *Kubelet) syncPod(pod *api.BoundPod, dockerContainers dockertools.Docke
 			}
 			if api.IsPullAlways(container.ImagePullPolicy) ||
 				(api.IsPullIfNotPresent(container.ImagePullPolicy) && (!present || latest)) {
+				kl.pullLock.RLock()
 				if err := kl.dockerPuller.Pull(container.Image); err != nil {
 					if ref != nil {
 
 						record.Eventf(ref, "failed", "failed", "Failed to pull image %q", container.Image)
 					}
 					glog.Errorf("Failed to pull image %q: %v; skipping pod %q container %q.", container.Image, err, podFullName, container.Name)
+					kl.pullLock.RUnlock()
```

Comment:
Actually, the list with "dangling=true" only returns images with no tag, images that you have just pulled will be tagged, so they won't be returned in the query for images to GC.

So all you have to do is protect the query against the pull and it should work correctly.

There is a small race window wherein if you take a dependency on an existing layer that was previously untagged, it will get re-tagged between the listing of untagged images and the delete of the image.

That's the other race that @anguslees 's PR was fixing, it is not addressed in this PR.  We should address it, but it's a much narrower race window than this race.